### PR TITLE
AP_NMEA_Output: add msg sentence $PASHR + refactor

### PR DESCRIPTION
**Pull request by magicrub**
add msg sentence $PASHR. This allows PhaseOne cameras to get orientation data to embed into it's images

### CODE REVIEW

**Review by bugeats — APPROVED** [review]
I can confirm this branch sends correct orientation data to the Phase One ixm-100 camera when using the simulator with an RS-232 adapter.

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (rmackay9) [inline libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:207]: This free looks incorrect. I guess it should be pashr or we'd get a memory leak.  There's an [issue](https://github.com/ArduPilot/ardupilot/issues/12514) and [PR](https://github.com/ArduPilot/ardupilot/pull/22184) which suggests we shouldn't be allocating and freeing all the time.

**Comment by rmackay9** [comment]
It looks like a useful change to me!

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (peterbarker) [inline libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:175]: Is this correct according to spec or should the field be empty or something?

**Review by peterbarker — COMMENTED** [review]
I wonder if we need a bitmask for sending this stuff....

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:198]: I know, right?

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:207]: I agree with not constantly allocating, bad practice for our embedded stuff. Meanwhile, this frees the already allocated memory. If we get here then pashr_res is -1 which means it failed to allocate pashr. See example of GPRMC above this code

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:190]: Whatever satisfies this....
`// 0 = All SPAN Pre-Alignment INS Status`

`// 1 = All SPAN Post-Alignment INS Status`

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:175]: I don't know, your guess is as good as mine. Safer to put it in than not?

**Comment by magicrub** [comment]
> I wonder if we need a bitmask for sending this stuff....

you know.. I had that all written up where it had a param enable bitmask and its own first-class library spot in AP_Vehicle... but trashed it to make sure this actually has any chance in hell of merging in any reasonable time. It also caused code bloat.

File: libraries/AP_NMEA_Output/AP_NMEA_Output.cpp
Comment (rmackay9) [reply libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:207]: Ah, ok, I see.

**Comment by rmackay9** [comment]
I also support adding a bitmask to allow choosing what data is output in the NMEA stream.  Boat users have requested that depth information is also output... obviously that's out of scope for this PR but just saying that if that param was added we might use it on Rover/Boats.

**Comment by magicrub** [comment]
Ok, so I went a little crazy on this. As requested, I've added a param for a bitmask but then it turned into more. Here's what this branch is now:
- single (first) commit adds PASHR msg (adds 456 bytes on CubeOrange, nothing on fmuv2)
- moved AP_NMEA_Output library to a first-class library to live in AP_Vehicle instead of AP_AHRS
- refactored the whole thing to utilize the string parsing and CheckSum generation code in AP_Common/NMEA.h
- Added Param: NMEA_RATE_MS to adjust the speed. Default is 100 (10Hz).
- Added Param: NMEA_MSG_EN bitmask to enable each of the three msgs types we support
- Added support for AP_Peirph (requires serial/GCS and GPS)


Adding PASHR increased codesize by 456 on CubeOrange, no change to fmuv2
Adding everything else, including the new PASHR msg, added a total of CubeOrange:360 bytes, fmuv2:0, CubeOrange-periph-heavy:1472

**Comment by magicrub** [comment]
@rmackay9 @WickedShell  So, we can bring in that single PASHR commit and make a new PR for the other stuff, or just keep them all together here. Your call(s).

File: libraries/AP_NMEA_Output/AP_NMEA_Output_config.h
Comment (peterbarker) [inline libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: `HAL_GCS_ENABLED` should always be defined?

File: libraries/AP_NMEA_Output/AP_NMEA_Output_config.h
Comment (peterbarker) [inline libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:9]: Would be nice to put the HAL_MINIMIZE_FEATURES into `minimal.inc`

File: libraries/AP_NMEA_Output/AP_NMEA_Output_config.h
Comment (peterbarker) [inline libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: Could we do all of this down in AP_Periph instead?

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_NMEA_Output/AP_NMEA_Output_config.h
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: fixed

Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: yes, but why?

Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: To move it to periph, the only way would be in AP_Peirph.h, and anyone else touching NMEA_Output, to include AP_NMEA_Output_config.h, then check if NMEA is enabled before including the normal AP_NMEA_Output.h.

Isn't the purpose of the _config to do this kind of stuff?

Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: @peterbarker And here's what that looks like in Periph.h
```
#include <AP_NMEA_Output/AP_NMEA_Output_config.h>
#if HAL_NMEA_OUTPUT_ENABLED
    #if HAL_GCS_ENABLED && defined(HAL_PERIPH_ENABLE_GPS)
        #include <AP_NMEA_Output/AP_NMEA_Output.h>
    #else
        // Needs SerialManager + (AHRS or GPS)
        #error "AP_NMEA_Output requires Serial/GCS and either AHRS or GPS. Needs HAL_GCS_ENABLED and HAL_PERIPH_ENABLE_GPS"
    #endif
#endif // HAL_NMEA_OUTPUT_ENABLED
```

**Comment by magicrub** [comment]
One major thing I just noticed on this is it now enables ```ALLOW_DOUBLE_MATH_FUNCTIONS``` on all builds. That was there before too but now it hits all the periphs and makes them 64bit maths even if the feature is disabled

File: libraries/AP_NMEA_Output/AP_NMEA_Output_config.h
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:6]: ok, I gave in. I moved it to periph

**Comment by magicrub** [comment]
looks like the allow_double_maths is probably fine

**Comment by magicrub** [comment]
I've tested this on valgrind with a TCP loopback.
- GPS1: Serial in via sim uBlox (truth data)
- GPS2: Type = 5, NMEA
- Serial5_protocol = 20 (NMEA Out). Loopback pointing to my GPS2 IP port

GPS1 and GPS2 tracked well. It was nice to see GPS2 lag (stair step changes) as I slowed down (NMEA_RATE_MS = 1000) and have it match almost perfectly as I sped up the rate (NMEA_RATE_MS = 50).

File: libraries/AP_NMEA_Output/AP_NMEA_Output_config.h
Comment (magicrub) [reply libraries/AP_NMEA_Output/AP_NMEA_Output_config.h]: in another PR...

**Comment by magicrub** [comment]
@peterbarker @tridge  OK to merge this?

**Comment by tridge** [comment]
@magicrub I'm happy with it, need approval from @peterbarker

**Review by peterbarker — APPROVED** [review]

**Pull request merged**
